### PR TITLE
Pass through based on context refactor

### PR DESCRIPTION
**Pull request by emmjohnson**
[PR requirements]: https://github.com/vmware-tanzu/cartographer/blob/main/CONTRIBUTING.md#commit-message-and-pr-guidelines

## Changes proposed by this PR

closes # 

## Release Note

[imperative mood]: https://chris.beams.io/posts/git-commit/#imperative

## PR Checklist

Note: Please do not remove items. Mark items as done `[x]` or use ~strikethrough~ if you believe they are not relevant

- [ ] Linked to a relevant issue. Eg: `Fixes #123` or `Updates #123`
- [ ] Removed non-atomic or `wip` commits
- [ ] Filled in the [Release Note](#Release-Note) section above 
- [ ] Modified the docs to match changes

### CODE REVIEW

File: pkg/apis/v1alpha1/cluster_delivery_webhook_test.go [hunk 1]
```diff
@@ -534,6 +534,80 @@ var _ = Describe("Delivery Validation", func() {
...
+					"error validating clusterdelivery [responsible-ops---default-params]: error validating resource [source-provider]: exactly one of option.Name or option.PassThrough must be specified, found neither",
+				))
+			})
+
+			It("deletes without error", func() {
+				Expect(delivery.ValidateDelete()).NotTo(HaveOccurred())
+			})
+		})
+
+		Context("options with pass through", func() {
```
Comment (idoru) [inline pkg/apis/v1alpha1/cluster_delivery_webhook_test.go:560]: No happy case test?

File: pkg/apis/v1alpha1/cluster_supply_chain_webhook_test.go [hunk 2]
```diff
@@ -976,5 +976,79 @@ var _ = Describe("Webhook Validation", func() {
...
+					"error validating clustersupplychain [responsible-ops---default-params]: error validating resource [source-provider]: exactly one of option.Name or option.PassThrough must be specified, found neither",
+				))
+			})
+
+			It("deletes without error", func() {
+				Expect(supplyChain.ValidateDelete()).NotTo(HaveOccurred())
+			})
+		})
+
+		Context("options with pass through", func() {
```
Comment (idoru) [inline pkg/apis/v1alpha1/cluster_supply_chain_webhook_test.go:1002]: ditto.

File: pkg/apis/v1alpha1/common.go [hunk 3]
```diff
@@ -162,8 +162,13 @@ func GetAPITemplate(templateKind string) (client.Object, error) {
...
 type TemplateOption struct {
 	// Name of the template to apply
+	// Name or PassThrough must be specified
 	// +kubebuilder:validation:MinLength=1
-	Name string `json:"name"`
+	Name string `json:"name,omitempty"`
+
+	// PassThrough the input
+	// Name or PassThrough must be specified
+	PassThrough string `json:"passThrough,omitempty"`
```
Comment (idoru) [inline pkg/apis/v1alpha1/common.go:171]: personal grype: I understand it's two words, but something about `passThrough` hurts my eyes.

File: pkg/realizer/component.go [hunk 4]
```diff
@@ -77,69 +77,92 @@ func (r *resourceRealizer) Do(ctx context.Context, resource OwnerResource, bluep
...
-		return nil, nil, nil, errors.GetTemplateError{
-			Err:           err,
-			ResourceName:  resource.Name,
-			TemplateName:  templateName,
-			BlueprintName: blueprintName,
-			BlueprintType: errors.SupplyChain,
+		var err error
+		stampReader, err = stamp.NewPassThroughReader(resource.TemplateRef.Kind, templateOption.PassThrough, inputGenerator)
+		if err != nil {
+			panic("whoops")
```
Comment (idoru) [inline pkg/realizer/component.go]: missed a panic

File: pkg/apis/v1alpha1/common_validations.go [hunk 5]
```diff
@@ -27,19 +27,31 @@ import (
...
+
 		if err := validateSelector(option.Selector, validPaths, validPrefixes); err != nil {
 			return fmt.Errorf("error validating option [%s] selector: %w", option.Name, err)
 		}
 	}
 
 	for _, option1 := range options {
+		name1 := option1.Name
+		if name1 == "" {
+			name1 = "passThrough"
```
Comment (idoru) [inline pkg/apis/v1alpha1/common_validations.go:61]: This seems to have the effect of only allowing one `passThrough` but I believe would generate a dubious error (saying you've got a duplicate) if one were to, say, specify something like:
```
        options:
          - passThrough: source-from-apples
            selector:
              matchExpressions:
                - { key: favorite-juice, operator: In, values: [ "apples" ] }
          - passThrough: source-from-oranges
            selector:
              matchExpressions:
                - { key: favorite-juice, operator: In, values: [ "apples" ] }
```

Should multiple passthroughs be allowed? If not, should error more specifically?

**Review by idoru — COMMENTED** [review]
barring comments, lgtm

File: pkg/apis/v1alpha1/common.go (hunk 3 above)
Comment (emmjohnson) [reply pkg/apis/v1alpha1/common.go:171]: I agree, but the spell checker squiggly line hurt my eyes more

File: pkg/apis/v1alpha1/common_validations.go (hunk 5 above)
Comment (emmjohnson) [reply pkg/apis/v1alpha1/common_validations.go:61]: Added additional validation to ensure only one or less pass throughs set

**Review by idoru — APPROVED** [review]
LGTM. I wonder if it is worth increasing the webhook validation scope to search for the matching source/image/config to fail faster?

File: pkg/apis/v1alpha1/cluster_delivery_webhook_test.go [hunk 6]
```diff
@@ -534,6 +534,160 @@ var _ = Describe("Delivery Validation", func() {
...
+					"error validating clusterdelivery [responsible-ops---default-params]: error validating resource [source-provider]: exactly one of option.Name or option.PassThrough must be specified, found neither",
+				))
+			})
+
+			It("deletes without error", func() {
+				Expect(delivery.ValidateDelete()).NotTo(HaveOccurred())
+			})
+		})
+
+		FContext("options with pass through", func() {
```
Comment (idoru) [inline pkg/apis/v1alpha1/cluster_delivery_webhook_test.go]:
```suggestion
		Context("options with pass through", func() {
```

**Review by idoru — APPROVED** [review]
LGTM after the FContext removal

**Pull request merged**